accepted/0042-isolated-mode.md

```diff
@@ -0,0 +1,347 @@
+
+### Duplication
+
+The `hoisted-mode` is not always successful at sharing common dependencies. Conflicts in versions lead to packages being installed multiple times on disk. These conflicts become more frequent when the project scales. The frequency of these conflicts can be reduced by an effort to align dependencies' versions across the project.
+
+These duplications come with the following cost:
+
+- Performance degradation of the installation phase
+- Inefficient disk usage
+- Performance and memory cost at runtime as Node.js sees duplicated modules as two different modules.
+
+### Tools crawling the node_modules folders
+
+Certain tools implement their own module resolution algorithm instead of the one provided by Node.js. One of the motivation for a tool to implement its own resolution algorithm is that it can add more features to it.
+
+Crawling the `node_modules` folder is such a feature used by certain build tools. This feature means that the mere presence of a package in a `node_modules` folder will have an impact on the output of a workspace. This means that any modification of a project's `node_modules` folder is possibly a breaking change to every workspaces in this project, regardless of the dependency graph.
+
+For example, the TypeScript compiler by default includes to its compilation _every_ package in the folder `node_modules/@types` without any explicit import statement needed.
+
+## Rationale
+
+The [import-maps](https://github.com/WICG/import-maps) standard is a very good tool to communicate the dependency graph to Node.js. But because this standard is not yet implemented in Node.js, npm needs an alternative way to solve the problems stated earlier.
+
+The goal is to provide an accurate dependency graph to Node.js while still relying on Node.js current module resolution algorithm.
+
+We deem valuable to invest the time implementing the `isolated-mode` now knowing that import-maps are coming because most of the implementation will be re-usable to implement a `import-maps-mode`.
+
+There are already two competing solutions out there which implement a `isolated-mode` ([pnpm](https://pnpm.io) and [yarn](https://yarnpkg.com)). We decided to choose the pnpm approach for the following reasons:
+
+- _Works with current ecosystem_, the pnpm `isolated-mode` does not require any modification to Node.js or to the various build tools.
+- _Battle tested_, Microsoft has successfully used pnpm to manage large monorepo for years.
+- _Recommended by Node.js_, this approach is actually the recommended approach by [the Node.js documentation](https://nodejs.org/api/modules.html#modules_addenda_package_manager_tips).
```

Comment:
This link doesn’t go anywhere relevant - i was very shocked to read the implication that node made any official recommendation here. It’d be great to track down this content, or remove this bullet point if it can’t be located.

Reply:
the doc changed, this is the new link, I will update it, in the meantime the link is this: https://nodejs.org/api/modules.html#modules_package_manager_tips

Reply:
oof, ok thanks - it even suggests using NODE_PATH, which is SUPER deprecated and not supported at all in ESM.

Reply:
While the NODE_PATH advice may be outdated, the advice related symlinks is probably not, the reasons for writing this in the NodeJs doc (presumably) still stand.

Reply:
I suspect the reasons are that the author preferred this approach, and nobody reviewing the PR noticed that it’d be inappropriate for node to make such statements ¯\\\_(ツ)\_/¯

Reply:
I can confidently report that yes, the author preferred that approach. :joy:

https://github.com/nodejs/node/commit/ff39d413a3c1#diff-5ae798d70a8dbd897d490bd7b453d5731ad3b49e416def53ee2ef0d11ca320b3R274